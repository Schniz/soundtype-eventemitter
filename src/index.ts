type Listener<T, Event extends keyof T> = (
  value: T[Event]
) => void | Promise<void>;

export class EventEmitter<T extends object> {
  private readonly listeners: Partial<
    {
      [key in keyof T]: (Listener<T, key>)[];
    }
  > = {};

  countListeners(): number {
    let sum = 0;
    for (const value of Object.values(this.listeners) as unknown[][]) {
      sum += value.length;
    }
    return sum;
  }

  emit<Event extends keyof T>(event: Event, value: T[Event]) {
    const listeners = this.listeners[event];

    if (listeners) {
      listeners.forEach(listener => listener(value));
    }
  }

  addListener<Event extends keyof T>(
    event: Event,
    onChange: Listener<T, Event>
  ) {
    let listeners = this.listeners[event];

    if (!listeners) {
      listeners = [];
      this.listeners[event] = listeners;
    }

    listeners.push(onChange) - 1;

    return () => this.removeListener(event, onChange);
  }

  removeListener<Event extends keyof T>(
    event: Event,
    onChange: Listener<T, Event>
  ) {
    let listeners = this.listeners[event];
    if (listeners) {
      const index = listeners.indexOf(onChange);
      if (index < 0) return;
      listeners.splice(index, 1);
    }
  }

  once<Event extends keyof T>(event: Event, fn: Listener<T, Event>) {
    const onChange: Listener<T, Event> = value => {
      fn(value);
      this.removeListener(event, onChange);
    };
    this.addListener(event, onChange);
  }

  waitFor<Event extends keyof T>(event: Event) {
    return new Promise<T[Event]>(resolve => this.once(event, resolve));
  }
}
