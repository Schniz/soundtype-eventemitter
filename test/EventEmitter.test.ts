import { EventEmitter } from '../src/index';
import { getTypes } from 'infer-types';
import * as path from 'path';

test('wait for one', async () => {
  const ee = new EventEmitter<{ hello: string }>();
  const waiter = ee.waitFor('hello');
  ee.emit('hello', 'world');
  expect(await waiter).toBe('world');
  expect(ee.countListeners()).toBe(0);
});

test('accepts async functions', done => {
  const ee = new EventEmitter<{ hello: string }>();
  ee.once('hello', async value => {
    expect(value).toBe('world');
    done();
  });
  ee.emit('hello', 'world');
});

test('assures types', async () => {
  const filepath = path.join(__dirname, 'sample.ts');
  const types = getTypes(filepath);
  expect(types).toMatchSnapshot('sample.ts exports');
});
