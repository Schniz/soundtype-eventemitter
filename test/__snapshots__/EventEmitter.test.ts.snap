// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assures types: sample.ts exports 1`] = `
Object {
  "hello": "string",
  "user": "User",
  "world": "number",
}
`;
